modules/aruco/src/aruco.cpp

```diff
@@ -0,0 +1,1573 @@
+    for(int i = 0; i < (int)candidatesOut.size(); i++) {
+        _candidates.create(4, 1, CV_32FC2, i, true);
+        Mat m = _candidates.getMat(i);
+        for(int j = 0; j < 4; j++)
+            m.ptr< Vec2f >(0)[j] = candidatesOut[i][j];
+
+        _contours.create((int)contoursOut[i].size(), 1, CV_32SC2, i, true);
+        Mat c = _contours.getMat(i);
+        for(unsigned int j = 0; j < contoursOut[i].size(); j++)
+            c.ptr< Point2i >()[j] = contoursOut[i][j];
+    }
+}
+
+
+/**
+  * @brief Given an input image and candidate corners, extract the bits of the candidate, including
+  * the border bits
+  */
+static Mat _extractBits(InputArray _image, InputArray _corners, int markerSize,
+                        int markerBorderBits, int cellSize, double cellMarginRate,
+                        double minStdDevOtsu) {
+
+    CV_Assert(_image.getMat().channels() == 1);
+    CV_Assert(_corners.total() == 4);
+    CV_Assert(markerBorderBits > 0 && cellSize > 0 && cellMarginRate >= 0 && cellMarginRate <= 1);
+    CV_Assert(minStdDevOtsu >= 0);
+
+    // number of bits in the marker
+    int markerSizeWithBorders = markerSize + 2 * markerBorderBits;
+    int cellMarginPixels = int(cellMarginRate * cellSize);
+
+    Mat resultImg; // marker image after removing perspective
+    int resultImgSize = markerSizeWithBorders * cellSize;
+    Mat resultImgCorners(4, 1, CV_32FC2);
+    resultImgCorners.ptr< Point2f >(0)[0] = Point2f(0, 0);
+    resultImgCorners.ptr< Point2f >(0)[1] = Point2f((float)resultImgSize - 1, 0);
+    resultImgCorners.ptr< Point2f >(0)[2] =
+        Point2f((float)resultImgSize - 1, (float)resultImgSize - 1);
+    resultImgCorners.ptr< Point2f >(0)[3] = Point2f(0, (float)resultImgSize - 1);
+
+    // remove perspective
+    Mat transformation = getPerspectiveTransform(_corners, resultImgCorners);
+    warpPerspective(_image, resultImg, transformation, Size(resultImgSize, resultImgSize),
+                    INTER_NEAREST);
+
+    // output image containing the bits
+    Mat bits(markerSizeWithBorders, markerSizeWithBorders, CV_8UC1, Scalar::all(0));
+
+    // check if standard deviation is enough to apply Otsu
+    // if not enough, it probably means all bits are the same color (black or white)
+    Mat mean, stddev;
+    // Remove some border just to avoid border noise from perspective transformation
+    Mat innerRegion = resultImg.colRange(cellSize / 2, resultImg.cols - cellSize / 2)
+                          .rowRange(cellSize / 2, resultImg.rows - cellSize / 2);
+    meanStdDev(innerRegion, mean, stddev);
+    if(stddev.ptr< double >(0)[0] < minStdDevOtsu) {
+        // all black or all white, depending on mean value
+        if(mean.ptr< double >(0)[0] > 127)
+            bits.setTo(1);
+        else
+            bits.setTo(0);
+        return bits;
+    }
+
+    // now extract code, first threshold using Otsu
+    threshold(resultImg, resultImg, 125, 255, THRESH_BINARY | THRESH_OTSU);
```

Comment:
125 and 255 are magic numbers. Can they be parameters or atleast named constants?